src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf16StringMarshaller.cs

```diff
@@ -84,15 +74,19 @@ public Utf16StringMarshaller(string? str, Span<ushort> buffer)
         /// <remarks>
         /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
         /// </remarks>
-        public void FromNativeValue(ushort* value) => _allocated = value;
+        public void FromNativeValue(ushort* value)
+        {
+            _nativeValue = value;
+            _allocated = true;
+        }
 
         /// <summary>
         /// Returns the managed string.
         /// </summary>
         /// <remarks>
         /// <seealso cref="CustomTypeMarshallerDirection.Out"/>
         /// </remarks>
-        public string? ToManaged() => _allocated == null ? null : new string((char*)_allocated);
+        public string? ToManaged() => Marshal.PtrToStringUni((IntPtr)_nativeValue);
```

Comment:
We should be consistent about whether the marshallers go through `Marshal.PtrToString` or perform the operation directly. There is observable behavior difference between the two options: The `Marshal` methods have special handling for the "Atom" values on Windows.

Reply:
Q: Is the current implementation just trying to match what the current built-in marshaling does? I guess we can live with that.

Reply:
> Is the current implementation just trying to match what the current built-in marshaling does?

Yes. We tried to model most of for .NET 7 off the current semantics. The plan was to iterate more once we have an understanding of usage and where the problems exist in the design. For example, we found some issues when aspnetcore was adopting it and that has pushed us to reconsider some restrictions we initially wanted to apply.

Reply:
> Is the current implementation just trying to match what the current built-in marshaling does?

Yes. We tried to model most of for .NET 7 off the current semantics. The plan was to iterate more once we have an understanding of usage and where the problems exist in the design. For example, we found some issues when aspnetcore was adopting it and that has pushed us to reconsider some restrictions we initially wanted to apply.